### ISSUE TITLE
Details passed to wamp.session.on_join subscribed asyncio method have bytes keys

### ISSUE THREAD

**Issue by terrycojones** [tavendo/AutobahnPython#388]
If I have a python2 router session that subscribes to `wamp.session.on_join`, the associated method is called with a details `dict` whose keys are strings.

However, I just converted the same code to be a python3 session and was surprised to get `KeyError`s. It turns out that the details passed to the method have `bytes` keys:

```
{b'authid': 'coordinator',
 b'authmethod': 'wampcra',
 b'authprovider': 'dynamic',
 b'authrole': 'coordinator',
 b'realm': 'realm1',
 b'session': 978859022,
 b'transport': {b'http_headers_received': {'cache-control': 'no-cache',
                                           'connection': 'Upgrade',
                                           'host': '127.0.0.1:8080',
                                           'pragma': 'no-cache',
                                           'sec-websocket-key': 'Zb3bDO+efx133r1EOMriQQ==',
                                           'sec-websocket-protocol': 'wamp.2.msgpack.batched,wamp.2.msgpack,wamp.2.json.batched,wamp.2.json',
                                           'sec-websocket-version': '13',
                                           'upgrade': 'WebSocket',
                                           'user-agent': 'AutobahnPython/0.10.3'},
                b'http_headers_sent': {},
                b'peer': b'tcp4:127.0.0.1:64774',
                b'protocol': b'wamp.2.msgpack.batched',
                b'type': b'websocket'}}
```

I could probably fix this, but I don't know where to look. Grepping for `wamp.session.on_join`  and `on_join` didn't turn up anything obvious.

**Comment by meejah** [comment]
This is a crossbar error, I created a [corresponding ticket](https://github.com/crossbario/crossbar/issues/316) there.

This should be closed.

**Comment by terrycojones** [comment]
Thanks @meejah.

**Comment by meejah** [comment]
I actually think there is an Autobahn bug here (see also the discusson on the crossbar bug): it's fairly common to use syntax like `dict(key='word')` which makes "plain" strings for the keys. When this goes through msgpack's packer with `use_bin_type=True`, the keys are then encoded as binary (which is against the WAMP standard).

Similarly, if you just took a `kwargs` dict and passed to Autobahn you'd have "plain" strings which become binary in msgpack (with `use_bin_type=True`)

Passing `use_bin_type=False` on Python2 would work to fix this, however that likely has other implications...?

**Comment by meejah** [comment]
One possible way to fix this and ensure we always have text keys for dicts is to use the msgpack  `object_hook` when deserializing and remap any non-text keys to text. There's no corresponding hook for serialization; we'd have to (recursively) inspect all incoming objects I guess.

See pull-request

**Comment by meejah** [comment]
There was still an issue with PR #390, now remedied with #395.

Unfortunately, it looks like v0.10.4 was tagged just after #390 was merged, so @oberstet unfortunately that means a v0.10.5 should be released with the #395 fix in it...Sorry about that :(

**Comment by oberstet** [comment]
@meejah Above addressed the immediate user issue. However, now I think this is pointing to a bigger issue: https://github.com/tavendo/AutobahnPython/issues/461. Apart from the performance impact the coercing of *all* dictionary keys to type string has.
